### PR TITLE
fixes(#814) allow plugins to extend the 'source' command group

### LINKED ISSUE #814: kn plugins for eventing sources should be able to extend `kn source` group
### Feature request

Current `kn` plugin architecture prevents plugins to extend existing command groups. This makes sense since we would not want a `kn-service-create` plugin that redefines the `service create` commands. Same for almost all other `kn` command groups.

But there is at least one exception. This is for command groups where we want to define sub-groups but at the same time allow extensions. One such example is the `kn source` eventing command group.

### Use case

Kn ships with some eventing sources, e.g., `ping`, but we also want to have plugins be created and defined that also extends this command group, e.g., `source kafka` and `source github` and others.

These plugins are being defined in the [client-contrib](https://github.com/knative/client-contrib) repo for instance the [`kn-source-kafka`](https://github.com/knative/client-contrib/pull/19) is one such example.

### UI Example

```
kn source kafka create my-kafka-source --sink my-kafka-sink ...
...
kn source github create my-hithub-source --sink my-github-sink
...
```

> /kind proposal
> /kind bug

### PR DESCRIPTION
**Pull request by maximilien**
## Description

In rare cases (e.g., `source`) an existing command group should be extensible via plugins. This PR addresses that. Source should be extensible since many of the eventing sources will be supported via plugins.

## Changes

* introduce a list of command group that can be extended (currently only source)
* check if plugin main group is in that list and execute, otherwise fail
* add e2e test for both plugin that is allowed and not allowed to extend existing group

## Reference

Fixes #814

/lint

### CODE REVIEW

**Comment by maximilien** [comment]
> tested the PR and seeing unknown command 'source'
> 
> ```
> ➜  client git:(7b488dfe) ./hack/build.sh -f
> 🚧 Compile
> 
> ➜  client git:(7b488dfe) ./kn plugin list
> /home/nshaikh/.config/kn/plugins/kn-source-test.sh
> ERROR:
>   - kn-source-test.sh overwrites existing built-in command: kn source
> kn-source-test.sh overwrites existing built-in command: kn source
> 
> ➜  client git:(7b488dfe) ./kn source test        
> Error: unknown command 'source' 
> Run 'kn --help' for usage.
> ```

Hmm, OK, thx, let me double check. Not seeing this on my side.

**Comment by maximilien** [comment]
OK @navidshaikh back on this. I think I see what's going on, I need to add more code to the `kn plugin list` which runs the verifier which looks for plugins extending all existing commands. However, the current code will actually allow you to create a plugin extending `source` command group. See following, which you can try as well.

```bash
➜  client git:(issue814) ls -la ~/.config/kn/plugins
total 87240
drwxr-xr-x  5 maximilien  staff       160 Apr 28 15:24 .
drwxr-xr-x  3 maximilien  staff        96 Apr  6 19:03 ..
-rwxr-xr-x  1 maximilien  staff       804 Mar 17 16:57 kn-curl
-rwxr-xr-x  1 maximilien  staff        35 Apr 28 15:24 kn-source-create
-rwx------  1 maximilien  staff  44657756 Apr 23 18:23 kn-source_github
➜  client git:(issue814) cat ~/.config/kn/plugins/kn-source-create
#!/bin/bash
echo "Hello kn: $1 $2"
➜  client git:(issue814) ./kn source create max
Hello kn: max
```
Let me try and add a fix for `kn plugin list` and perhaps more tests.

**Comment by maximilien** [comment]
@navidshaikh see latest updates. Also did a series of local tests that worked well for me:

```bash
➜  client git:(issue814) ✗ ls -la ~/.config/kn/plugins
total 87256
drwxr-xr-x  7 maximilien  staff       224 Apr 28 15:49 .
drwxr-xr-x  3 maximilien  staff        96 Apr  6 19:03 ..
-rwxr-xr-x  1 maximilien  staff       804 Mar 17 16:57 kn-curl
-rwxr-xr-x  1 maximilien  staff        35 Apr 28 15:49 kn-service-create
-rwxr-xr-x  1 maximilien  staff        35 Apr 28 15:49 kn-service-max
-rwxr-xr-x  1 maximilien  staff        35 Apr 28 15:24 kn-source-create
-rwx------  1 maximilien  staff  44657756 Apr 23 18:23 kn-source_github
➜  client git:(issue814) ✗ ./kn plugin list
/Users/maximilien/.config/kn/plugins/kn-curl
/Users/maximilien/.config/kn/plugins/kn-service-create
/Users/maximilien/.config/kn/plugins/kn-service-max
/Users/maximilien/.config/kn/plugins/kn-source-create
/Users/maximilien/.config/kn/plugins/kn-source_github
ERRORs:
  - kn-service-create overwrites existing built-in command: kn service create
  - kn-service-max overwrites existing built-in command: kn service
kn-service-create overwrites existing built-in command: kn service create,kn-service-max overwrites existing built-in command: kn service
➜  client git:(issue814) ✗ ./kn service max
Error: unknown subcommand 'max' for 'kn service'. Available subcommands: create, delete, describe, export, list, update
Run 'kn --help' for usage.
➜  client git:(issue814) ✗ ./kn source create max
Hello kn: max
➜  client git:(issue814) rm ~/.config/kn/plugins/kn-service-*
➜  client git:(issue814) ls -la ~/.config/kn/plugins
total 87240
drwxr-xr-x  5 maximilien  staff       160 Apr 28 15:59 .
drwxr-xr-x  3 maximilien  staff        96 Apr  6 19:03 ..
-rwxr-xr-x  1 maximilien  staff       804 Mar 17 16:57 kn-curl
-rwxr-xr-x  1 maximilien  staff        35 Apr 28 15:24 kn-source-create
-rwx------  1 maximilien  staff  44657756 Apr 23 18:23 kn-source_github
➜  client git:(issue814) ./kn plugin list
/Users/maximilien/.config/kn/plugins/kn-curl
/Users/maximilien/.config/kn/plugins/kn-source-create
/Users/maximilien/.config/kn/plugins/kn-source_github
➜  client git:(issue814) ./kn source create max
Hello kn: max
```
Hopefully we can merge this as @daisy-ycguo and I need this feature for our `kn-source-kafka` and `kn-source-github` plugins and the many other source plugins that will follow.

**Comment by maximilien** [comment]
Hi, @navidshaikh, we should be good to go now. Please merge when greeen.

cc @daisy-ycguo

File: pkg/kn/commands/plugin/list_test.go [hunk 1]
```diff
@@ -201,6 +199,24 @@ func TestPluginList(t *testing.T) {
...
+							Use: "source",
+						}
+						ctx.rootCmd.AddCommand(sourceCmd)
+						defer ctx.rootCmd.RemoveCommand(sourceCmd)
+
+						err := ctx.createTestPlugin("kn-source-fake", FileModeExecutable, true)
+						assert.NilError(t, err)
+
+						err = ctx.execute("plugin", "list", "--lookup-plugins=true")
+						assert.NilError(t, err)
```
Comment (navidshaikh) [inline pkg/kn/commands/plugin/list_test.go:217]: should this also verify the plugin name from the returned list ?

File: pkg/kn/commands/plugin/plugin.go [hunk 2]
```diff
@@ -52,3 +52,20 @@ func BindPluginsFlagToViper(cmd *cobra.Command) {
 	viper.SetDefault("plugins-dir", commands.Cfg.DefaultPluginDir)
 	viper.SetDefault("lookup-plugins", false)
 }
+
+// AllowedExtensibleCommandGroups the list of command groups that can be
+// extended with plugins, e.g., a plugin named `kn-source-kafka` for Kafka
+// event sources is allowed. This is defined as a fixed [...]string since
+// cannot defined Golang []string constants
+var AllowedExtensibleCommandGroups = [...]string{"source"}
```
Comment (navidshaikh) [inline pkg/kn/commands/plugin/plugin.go:60]: I think the command group to extend should be configurable outside the kn, i.e config.
OR
It shouldn't mind extending command groups if the plugin command dont clash.

Comment (maximilien) [reply pkg/kn/commands/plugin/plugin.go:60]: Not sure about that since we would not want extensions of core commands

Comment (maximilien) [reply pkg/kn/commands/plugin/plugin.go:60]: as per today’s discussion, please merge this PR and I will open an issue with your suggestion with more details and open a 24 hours vote. Thanks @navidshaikh and @rhuss

**Pull request merged**